sydent/threepid/bind.py

```diff
@@ -101,29 +102,26 @@ def _notify(self, assoc, attempt):
         mxid = assoc["mxid"]
         domain = mxid.split(":")[-1]
         server = yield self._pickServer(domain)
-        callbackUrl = "https://%s/_matrix/federation/v1/3pid/onbind" % (
+
+        post_url = "https://%s/_matrix/federation/v1/3pid/onbind" % (
             server,
         )
 
-        logger.info("Making bind callback to: %s", callbackUrl)
-        # TODO: Not be woefully insecure
-        agent = Agent(reactor, InsecureInterceptableContextFactory())
-        reqDeferred = agent.request(
-            "POST",
-            callbackUrl.encode("utf8"),
-            Headers({
-                "Content-Type": ["application/json"],
-                "User-Agent": ["Sydent"],
-            }),
-            FileBodyProducer(StringIO(json.dumps(assoc)))
-        )
-        reqDeferred.addCallback(
-            lambda _: logger.info("Successfully notified on bind for %s" % (mxid,))
-        )
+        logger.info("Making bind callback to: %s", post_url)
 
-        reqDeferred.addErrback(
-            lambda err: self._notifyErrback(assoc, attempt, err)
-        )
+        # Make a POST to the chosen Synapse server
+        http_client = FederationHttpClient(self.sydent)
+        try:
+            response = yield http_client.post_json_get_nothing(post_url, assoc, {})
+            # If the request failed, try again with exponential backoff
+            if response.code != 200:
+                self._notifyErrback(
+                    assoc, attempt, "Non-OK error code received (%d)" % response.code
+                )
+            else:
+                logger.info("Successfully notified on bind for %s" % (mxid,))
+        except Exception as e:
+            self._notifyErrback(assoc, attempt, e)
```

Comment:
this will get called if the first `_notifyErrback` raises, which is admittedly unlikely, but it would be nice to make it right. How about:

```
        try:
            response = yield http_client.post_json_get_nothing(post_url, assoc, {})
        except Exception as e:
            self._notifyErrback(assoc, attempt, e)
            return

        # now we can process response
````